### ISSUE TITLE
[REC] Ret: Seraphim, Final Reckoning and Divine Toll get out of sync in AoE

### ISSUE THREAD

**Issue by cremor** [Hekili/hekili#1422]
### Before You Begin

- [X] I confirm that I have downloaded the latest version of the addon.
- [X] I am not playing on a private server.
- [X] I checked for an [existing, open ticket](https://github.com/Hekili/hekili/labels/recommendation) for this issue and was not able to find one.
- [X] I edited the title of this issue (above) so that it describes the issue I am reporting.
- [X] I am reporting an issue with the default priority included with the specialization (imported or edited priorities are not supported).

### Describe the Issue

When you play Ret as Kyrian, use Mikanikos as soulbind and both Seraphim and Final Reckoning talents, the current recommendations will causes a desync of Seraphim, Final Reckoning and Divine Toll in AoE situations.

On the pull everything is still fine. But if you later use Seraphim when it's recommended, neither Final Reckoning nor Divine Toll will be shown again for the duration of the fight.

The problem only happens when cooldowns are disabled or shown separately.

The cause seems to be the cooldown reduction of Divine Toll granted by Mikanikos when it hits multiple targets. This causes the Seraphim recommendation to show up earlier than it would be on single target. And when Final Reckoning then comes off cooldown, the Seraphim buff is alreay gone, so Final Reckoning will not be recommended.
Divine Toll isn't recommended either because "generators 5" only shows it without the Final Reckoning talent and the "es_fr_active 3" will not be shown because "es_fr_active" isn't shown outside of the Final Reckoning window.

### How to Reproduce

1. Play as Kyrian Ret Paladin with Mikanikos soulbind and Seraphim and Final Reckoning talents.
2. Either disable cooldowns in Hekili or configure them to show separately.
3. Start an AoE fight and follow the recommendations until Seraphim is usable the second time.
4. Use Seraphim as recommended.
5. Notice that Divine Toll will not be recommended in the Seraphim buff window.
6. Notice that Final Reckoning will not be recommended when it's off cooldown again.

### Snapshot (Link)

https://pastebin.com/jc8fAAhT

### Raidbots Sim Report (Link)

https://www.raidbots.com/simbot/report/fKJnxnrMBnsyGt86bFB6Nq

**Comment by Hekili** [comment]
> But if you later use Seraphim when it's recommended, neither Final Reckoning nor Divine Toll will be shown again for the duration of the fight.

Yeah, having Seraphim fully intended to sync with cooldowns, but not being a cooldown itself, and then splitting it into a separate window breaks a lot of logic.  In the Primary display, Divine Toll's actual CD is used, but Final Reckoning's isn't.

> The problem only happens when cooldowns are disabled or shown separately.

Every time I see this, I'm more inclined to delete the separate Cooldowns window, because priorities are just not written to split abilities out in a non-linear way.

Anyway, I've made adjustments in 1.0.9-alpha2 that you can try out.  Things to try:
- [ ] Just see if everything is working without making any changes.
- [ ] If Cooldowns: Show Separately is enabled, go to /hek > Retribution > Toggles.  In the Cooldowns section, check the new `Cooldowns: Show Separately - Use Actual Cooldowns` feature and see if that helps.
- [ ] Any other testing/tweaking you want to try and report back

Thanks!

**Comment by cremor** [comment]
>  Just see if everything is working without making any changes.

Looks far better now, thanks!
The only small remaining problem I found is that it can take quite a while for Final Reckoning to show up as a separate cooldown, even if it's already available. I assume this is because Seraphim and Divine Toll are used earlier (not synced with separately shown cooldowns) and the Seraphim buff isn't up any more.
Here is a snapshot: https://pastebin.com/Rf46X11F

Anyway, I don't think that this is a big problem. Using it at the point in time of the snapshot would also cause the next Seraphim/Divine Toll usage to not be in sync any more, so maybe it might even be a DPS loss.

> If Cooldowns: Show Separately is enabled, go to /hek > Retribution > Toggles. In the Cooldowns section, check the new Cooldowns: Show Separately - Use Actual Cooldowns feature and see if that helps.

This looks promising, it keeps the abilities synced up correctly in AoE. But it has it's own problems. If you need to hold your CDs, you'll end up staying in "es_fr_pooling". For example, in this snapshot I'm at 5 HP but it is recommended that I wait for Consecration (which is still on cooldown), although using a finisher would be better normally when not using CDs soon:
https://pastebin.com/bayMBnP9

> Any other testing/tweaking you want to try and report back

I've only tested on target dummies for now. One thing I need to keep watching are recommendations in M+ after the CD reduction buff from the new affix is gained. I'm not sure if there are problems with that buff, but at least I had a few desyncs yesterday (before using the new alpha) in M+, even though I had cooldown enabled and not shown separately.

Btw, because of all that cooldown dependend recommendations I've now personally disabled "show separately" and bound a key for the cooldowns toggle so that I can quickly disable them when I don't want to use CDs yet. But with that new "Use Actual Cooldowns" feature I might consider switching to "show separately" again if it can be optimized to not have downsides any more (or when it can be quickly toggled with a keybind).

**Comment by Hekili** [comment]
> The only small remaining problem I found is that it can take quite a while for Final Reckoning to show up as a separate cooldown, even if it's already available. I assume this is because Seraphim and Divine Toll are used earlier (not synced with separately shown cooldowns) and the Seraphim buff isn't up any more.
> Here is a snapshot: [pastebin.com/Rf46X11F](https://pastebin.com/Rf46X11F)

It's because Seraphim is desynched and the entry for Final Reckoning in the Cooldowns list wants Seraphim to be active before using it.  I can't tell what Seraphim's cooldown is in that snapshot because it looks disabled.  It's possible to add logic that sets a threshold where Final Reckoning would just be used without Seraphim, but it would be challenging to tune that timeframe precisely.  Is it better to wait as much as 30 seconds to sync up Final Reckoning with Seraphim?  Or better to desync twice (by 15 seconds) to line them up on the 3rd.  I don't know, and am not sure how to test.

> This looks promising, it keeps the abilities synced up correctly in AoE. But it has it's own problems. If you need to hold your CDs, you'll end up staying in "es_fr_pooling". For example, in this snapshot I'm at 5 HP but it is recommended that I wait for Consecration (which is still on cooldown), although using a finisher would be better normally when not using CDs soon:  [pastebin.com/bayMBnP9](https://pastebin.com/bayMBnP9)

Yes, the assumption is that, when you have Cooldowns enabled, you'll use them and not sit on them.  If you don't want the Primary display to make that assumption, you have to toggle them off, even when they're in a separate window.

The impact of the new setting is that it *doesn't* fake those abilities being on cooldown so that, when cooldowns are enabled, the Primary recommendations assume you're going to push whatever is in the Cooldowns display when it pops up.  Again, not a fan of showing Cooldowns in a separate display for exactly this reason -- it's fiddly and almost nobody understands the limitations of it.

However, in that snapshot it's showing cooldowns enabled, it's not showing cooldowns shown separately, but it's saying Final Reckoning is toggled off.  Did you move FR to a different toggle?

> I've only tested on target dummies for now. One thing I need to keep watching are recommendations in M+ after the CD reduction buff from the new affix is gained. I'm not sure if there are problems with that buff, but at least I had a few desyncs yesterday (before using the new alpha) in M+, even though I had cooldown enabled and not shown separately.

The whole priority is written to sync CDs, so CDR effects that break that synchronization aren't generally accounted for.  Every place the addon's priority accounts for it is a place I've explicitly written that in.  I have to add it on a case-by-case basis, but don't have a way to test in sims to make sure it's worth it.  It's not what I want to be doing with the addon, but I also don't want to leave people hanging in M+ without support.

> Btw, because of all that cooldown dependend recommendations I've now personally disabled "show separately" and bound a key for the cooldowns toggle so that I can quickly disable them when I don't want to use CDs yet. But with that new "Use Actual Cooldowns" feature I might consider switching to "show separately" again if it can be optimized to not have downsides any more (or when it can be quickly toggled with a keybind).

Using the toggle (without the new feature turned on) is the safe and consistent way, which is why it's been available for 7 or 8 years.  It makes it so that the addon isn't trying to prepare for you to use CDs while you're not using them, and that addon logic works normally when you turn CDs on.

**Comment by cremor** [comment]
>  I can't tell what Seraphim's cooldown is in that snapshot because it looks disabled.

It was not disabled. If I remember correctly, I used it about 15-20 seconds before that snapshot was created. It was recommended by the primary display at that time.
Note that that snapshot is from the cooldowns display. So I assume it treats Seraphim as disabled because it's not configured to be a cooldown?

>  It's possible to add logic that sets a threshold where Final Reckoning would just be used without Seraphim, but it would be challenging to tune that timeframe precisely. Is it better to wait as much as 30 seconds to sync up Final Reckoning with Seraphim? Or better to desync twice (by 15 seconds) to line them up on the 3rd. I don't know, and am not sure how to test.

Yeah, I don't know either. Maybe it's not justified to invest more time in this. Who knows what the next problem will be if we go down this rabbit hole.

> Yes, the assumption is that, when you have Cooldowns enabled, you'll use them and not sit on them. If you don't want the Primary display to make that assumption, you have to toggle them off, even when they're in a separate window.
[...]
Again, not a fan of showing Cooldowns in a separate display for exactly this reason -- it's fiddly and almost nobody understands the limitations of it.

I understand that now, but it took some time. For a long time this didn't matter for me - either the APL didn't care about remaining timers on cooldowns for recommending "normal" abilites before, or I just used talent combinations were it didn't care (e.g. I just started playing with Final Reckoning).

There is already a warning in the tooltip for "show separately", but maybe it needs to be clearified? E.g. including something like "the assumption is that, when you have Cooldowns enabled, you'll use them and not sit on them" would be good.

> However, in that snapshot it's showing cooldowns enabled, it's not showing cooldowns shown separately, but it's saying Final Reckoning is toggled off. Did you move FR to a different toggle?

That should have been the snapshot with the new option (and CDs shown separately). Unless I copied the wrong snapshot... But the line "Recommendation # 1 is consecration at 1.50s (1.50s)." indicates that this is the snapshot of what I saw.
Is the "separate" setting even written to the snapshot? I can't see it in my old snapshots either.

Final Reckoning is enabled and on the toggle "Default (cooldowns)".

> It's not what I want to be doing with the addon, but I also don't want to leave people hanging in M+ without support.

I understand. I'll keep an eye on it, but it's really hard to tell in the middle of an M+ whether a desync was caused by the new CD reduction buff or by me not following the recommendations 😀
